lib/ansible/playbook/role/__init__.py

```diff
@@ -460,17 +472,30 @@ def get_dep_chain(self):
             dep_chain.append(parent)
         return dep_chain
 
+    @property
+    def cached_deps(self):
+        return global_role_dependency_cache.get(self.full_qualified_name, None)
+
+    def add_deps_to_cache(self, deps):
+        global_role_dependency_cache[self.full_qualified_name] = deps
+
     def get_default_vars(self, dep_chain=None):
         dep_chain = [] if dep_chain is None else dep_chain
 
-        default_vars = dict()
-        for dep in self.get_all_dependencies():
-            default_vars = combine_vars(default_vars, dep.get_default_vars())
-        if dep_chain:
-            for parent in dep_chain:
-                default_vars = combine_vars(default_vars, parent._default_vars)
-        default_vars = combine_vars(default_vars, self._default_vars)
-        return default_vars
+        deps = self.cached_deps
+        if deps is None:
+            deps = self.get_all_dependencies()
+            self.add_deps_to_cache(deps)
```

Comment:
I noticed that this block is repeated. It might be an opportunity to turn it into a `.cached_deps` + rename the current `.cached_deps` into `.global_cached_deps`. Not sure. Something isn't feeling quite right, but I can't express it. Ideas?

Reply:
I think I found a good solution with my last commit (88279093bdc60f4d29d74340ff90b8e126bc1980). Could you please review whether this fits with what you had in mind?